File: content/snippets/python/interview_lcci_17_17_multi_search.py
from typing import List


class Trie:
    def __init__(self):
        self.children = [None] * 26
        self.idx = -1

    def insert(self, word: str, i: int):
        node = self
        for c in word:
            j = ord(c) - ord('a')
            if node.children[j] is None:
                node.children[j] = Trie()
            node = node.children[j]
        node.idx = i


class Solution:
    def multiSearch(self, big: str, smalls: List[str]) -> List[List[int]]:
        trie = Trie()
        for i, s in enumerate(smalls):
            if s:
                trie.insert(s, i)

        ans = [[] for _ in smalls]
        for i in range(len(big)):
            node = trie
            for j in range(i, len(big)):
                k = ord(big[j]) - ord('a')
                if node.children[k] is None:
                    break
                node = node.children[k]
                if node.idx != -1:
                    ans[node.idx].append(i)
        return ans

File: content/snippets/cpp/interview_lcci_17_17_multi_search.cpp
#include <string>
#include <vector>
using namespace std;

struct Trie {
    Trie* children[26] = {};
    int idx = -1;

    void insert(const string& word, int i) {
        Trie* node = this;
        for (char c : word) {
            int j = c - 'a';
            if (!node->children[j]) node->children[j] = new Trie();
            node = node->children[j];
        }
        node->idx = i;
    }
};

class Solution {
public:
    vector<vector<int>> multiSearch(string big, vector<string>& smalls) {
        Trie* trie = new Trie();
        for (int i = 0; i < (int)smalls.size(); ++i) {
            if (!smalls[i].empty()) trie->insert(smalls[i], i);
        }
        int n = big.size();
        vector<vector<int>> ans(smalls.size());
        for (int i = 0; i < n; ++i) {
            Trie* node = trie;
            for (int j = i; j < n; ++j) {
                int k = big[j] - 'a';
                if (!node->children[k]) break;
                node = node->children[k];
                if (node->idx != -1) ans[node->idx].push_back(i);
            }
        }
        return ans;
    }
};
